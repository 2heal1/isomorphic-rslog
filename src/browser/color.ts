let supportsSubstitutions: boolean | undefined = undefined;

export const supportColor = () => {
  if (typeof supportsSubstitutions !== 'undefined') {
    return supportsSubstitutions;
  }
  
  const originalConsoleLog = console.log;
  try {
    const testString = 'color test';
    const css = 'color: red;';
    supportsSubstitutions = false;
    
    console.log = (...args) => {
      if (args[0] === `%c${testString}` && args[1] === css) {
        supportsSubstitutions = true;
      }
    };
    
    console.log(`%c${testString}`, css);
  } catch (e) {
    supportsSubstitutions = false;
  } finally {
    console.log = originalConsoleLog;
  }
  
  return supportsSubstitutions;
};

export type ColorFn = (input: string | number | null | undefined | [label: string, style: string]) => string[];

const ansiToCss = {
  'bold': 'font-weight: bold;',
  'red': 'color: red;',
  'green': 'color: green;',
  'orange': 'color: orange;',
  'dodgerblue': 'color: dodgerblue;',
  'magenta': 'color: magenta;',
  'gray': 'color: gray;',
};

let formatter = (key: keyof typeof ansiToCss): ColorFn =>
  supportColor()
    ? (input) => {
      if (Array.isArray(input)) {
        const [label, style] = input;
        return [`%c${label.replace('%c', '')}`, style ? `${ansiToCss[key]}${style}` : `${ansiToCss[key] || ''}`];
      }
      return [`%c${String(input).replace('%c', '')}`, ansiToCss[key] || ''];
    }
    : (input) => [String(input)];


export const bold = formatter('bold');
export const red = formatter('red');
export const green = formatter('green');
export const orange = formatter('orange');
export const dodgerblue = formatter('dodgerblue');
export const magenta = formatter('magenta');
export const gray = formatter('gray');