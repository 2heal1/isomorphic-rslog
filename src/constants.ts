export let LOG_LEVEL = {
  error: 0,
  warn: 1,
  info: 2,
  log: 3,
  verbose: 4,
} as const;
