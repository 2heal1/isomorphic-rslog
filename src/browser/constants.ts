import type { LogType } from '../types';
import { red , yellow , cyan , green , magenta  } from './color';

export let LOG_TYPES = {
  // Level error
  error: {
    label: 'error',
    level: 'error',
    color:  red,
  },
  // Level warn
  warn: {
    label: 'warn',
    level: 'warn',
    color:  yellow,
  },
  // Level info
  info: {
    label: 'info',
    level: 'info',
    color:  cyan,
  },
  start: {
    label: 'start',
    level: 'info',
    color:  cyan,
  },
  ready: {
    label: 'ready',
    level: 'info',
    color:  green,
  },
  success: {
    label: 'success',
    level: 'info',
    color:  green,
  },
  // Level log
  log: {
    level: 'log',
  },
  // Level debug
  debug: {
    label: 'debug',
    level: 'verbose',
    color:  magenta,
  },
} satisfies Record<string, LogType>;
