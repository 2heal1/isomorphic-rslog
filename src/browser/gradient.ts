import { supportColor } from './color'
const startColor = [189, 255, 243]; // RGB for #bdfff3
const endColor = [74, 194, 154];    // RGB for #4ac29a

const isWord = (char: string) => !/[\s\n]/.test(char);

export function gradient(message: string): string[] {
  if (!supportColor()) {
    return [message]
  }
  const chars: string[] = [...message];
  const words = chars.filter(isWord);
  const steps = words.length - 1;

  if (steps === 0) {
    return [`%c${message}`, `color: rgb(${startColor.join(',')}); font-weight: bold;`];
  }

  let output = '';
  let styles: string[] = [];

  chars.forEach((char) => {
    if (isWord(char)) {
      const progress = words.indexOf(char) / steps;
      const r = Math.round(startColor[0] + (endColor[0] - startColor[0]) * progress);
      const g = Math.round(startColor[1] + (endColor[1] - startColor[1]) * progress);
      const b = Math.round(startColor[2] + (endColor[2] - startColor[2]) * progress);

      output += `%c${char}`;
      styles.push(`color: rgb(${r},${g},${b}); font-weight: bold;`);
    } else {
      output += char;
    }
  });

  return [output, ...styles]
}