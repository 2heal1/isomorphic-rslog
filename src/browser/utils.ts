import type{Labels, LogMessage, LogMethods, LogType} from '../types'
import { bold } from './color';

export function getLabel(type:LogMethods,logType: LogType,labels:Labels){
  let label = [''];
  if ('label' in logType) {
    label = [(labels[type] || logType.label || '')];
    label = bold(logType.color ? logType.color(label as [string,string]) as [string,string] : label[0]);
  }
  label = label.filter(Boolean)

  return label
}

export function finalLog(label:string[],text:string,args:string[],message?:LogMessage){
  if(label.length){
     //  gradient
    if(Array.isArray(message)){
      console.log(...label,...message)
    }else{
      console.log(...label,text)
    }
  }else{
    Array.isArray(message) ? console.log(...message) : console.log(text,...args)
  }
}