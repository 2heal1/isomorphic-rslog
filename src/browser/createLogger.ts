import { finalLog, getLabel } from './utils';
import { createLogger as rawCreateLogger } from '../createLogger';
import { gradient } from './gradient';
import type { Options, } from '../types';
import { LOG_TYPES } from './constants'

export function createLogger(options: Options = {},) {
   return rawCreateLogger(options, {
      handleError: (msg) => msg,
      getLabel,
      gradient,
      finalLog,
      LOG_TYPES,
      greet: (msg) => {
         return gradient(msg)
      }
   })
}