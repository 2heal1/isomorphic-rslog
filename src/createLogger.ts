import { LOG_LEVEL, } from './constants';
import { isErrorStackMessage } from './utils';
import type { Options, LogMessage, Logger, LogMethods, LogTypes } from './types';

import type { gradient as TGradient } from './browser/gradient';
import type { finalLog as TFinalLog, getLabel as TGetLabel } from './browser/utils';

function validateOptions(options: Options): Options {
  const validatedOptions = { ...options };
  if (options.labels && typeof options.labels !== 'object') {
    throw new Error('Labels must be an object');
  }
  if (options.level && typeof options.level !== 'string') {
    throw new Error('Level must be a string');
  }
  return validatedOptions;
}

export let createLogger = (options: Options = {}, { getLabel, handleError, finalLog, greet, LOG_TYPES }: {
  LOG_TYPES: LogTypes;
  getLabel: typeof TGetLabel;
  finalLog: typeof TFinalLog;
  gradient: typeof TGradient,
  greet: (msg: string) => unknown,
  handleError: (msg: string) => string,

}) => {
  const validatedOptions = validateOptions(options);
  let maxLevel = validatedOptions.level || 'log';
  let customLabels = validatedOptions.labels || {};

  let log = (type: LogMethods, message?: LogMessage, ...args: string[]) => {
    if (LOG_LEVEL[LOG_TYPES[type].level] > LOG_LEVEL[maxLevel]) {
      return;
    }

    if (message === undefined || message === null) {
      return console.log();
    }

    let logType = LOG_TYPES[type];
    let text = '';

    const label = getLabel(type, logType, customLabels);

    if (message instanceof Error) {
      if (message.stack) {
        let [name, ...rest] = message.stack.split('\n');
        if (name.startsWith('Error: ')) {
          name = name.slice(7);
        }
        text = `${name}\n${handleError(rest.join('\n'))}`;
      } else {
        text = message.message;
      }
    }

    // change the color of error stacks to
    else if (logType.level === 'error' && typeof message === 'string') {

      let lines = message.split('\n');
      text = lines
        .map(line => (isErrorStackMessage(line) ? handleError(line) : line))
        .join('\n');
    } else {
      text = `${message}`;
    }

    finalLog(label, text, args, message)
  };

  let logger = {
    // greet
    greet: (message: string) => log('log', greet(message)),
  } as Logger;

  (Object.keys(LOG_TYPES) as LogMethods[]).forEach(key => {
    logger[key] = (...args) => log(key, ...args);
  });

  Object.defineProperty(logger, 'level', {
    get: () => maxLevel,
    set(val) {
      maxLevel = val;
    },
  });

  Object.defineProperty(logger, 'labels', {
    get: () => customLabels,
    set(val) {
      customLabels = val;
    },
  });

  logger.override = customLogger => {
    Object.assign(logger, customLogger);
  };

  return logger;
};
