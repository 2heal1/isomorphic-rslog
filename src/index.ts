import { createLogger } from './createLogger';

export { createLogger } from './createLogger';

export const logger = createLogger();

export type {
  Options,
  LogType,
  LogLevel,
  LogMessage,
  LogFormatter,
  LogFunction,
} from './types';
