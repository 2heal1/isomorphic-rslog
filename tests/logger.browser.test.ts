import { createLogger, logger } from '../src/browser';
import { join } from 'path';
import {
  expect,
  test,
  describe,
  vi,
  Mock,
  beforeEach,
  afterEach,
} from 'vitest';

const root = join(__dirname, '../');

// Mock color module
vi.mock('../src/browser/color', () => {
  const colorFn = (input: string | [string, string]) => {
    if (Array.isArray(input)) {
      const [label, style] = input;
      return [`%c${label}`, style];
    }
    return [`%c${input}`, 'font-weight: bold;'];
  };

  const createColorFn = (color: string): typeof colorFn => {
    return input => {
      if (Array.isArray(input)) {
        const [label, style] = input;
        return [
          `%c${label}`,
          `font-weight: bold;color: ${color};${style ? ` ${style}` : ''}`,
        ];
      }
      return [`%c${input}`, `font-weight: bold;color: ${color};`];
    };
  };

  let supportsSubstitutions = true;

  return {
    supportColor: vi.fn(() => supportsSubstitutions),
    bold: colorFn,
    red: createColorFn('red'),
    green: createColorFn('green'),
    orange: createColorFn('orange'),
    dodgerblue: createColorFn('dodgerblue'),
    magenta: createColorFn('magenta'),
    gray: createColorFn('gray'),
  };
});

expect.addSnapshotSerializer({
  test: val => typeof val === 'string' && val.includes(root),
  print: val => {
    return (val as any).toString().replaceAll(root, '<ROOT>');
  },
});

describe('logger', () => {
  test('should log as expected', () => {
    console.log = vi.fn();
    logger.info('test info');
    logger.warn('test warn');
    logger.error('test error');

    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(3);

    // Info log
    expect(calls[0][0]).toBe('%c%cinfo');
    expect(calls[0][1]).toBe('font-weight: bold;color: dodgerblue;');
    expect(calls[0][2]).toBe('test info');

    // Warn log
    expect(calls[1][0]).toBe('%c%cwarn');
    expect(calls[1][1]).toBe('font-weight: bold;color: orange;');
    expect(calls[1][2]).toBe('test warn');

    // Error log
    expect(calls[2][0]).toBe('%c%cerror');
    expect(calls[2][1]).toBe('font-weight: bold;color: red;');
    expect(calls[2][2]).toBe('test error');
  });

  test('should create new logger with info level correctly', () => {
    console.log = vi.fn();
    const newLogger = createLogger({ level: 'info' });
    newLogger.info('test info');
    newLogger.warn('test warn');
    newLogger.error('test error');

    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(3);

    // Info log
    expect(calls[0][0]).toBe('%c%cinfo');
    expect(calls[0][1]).toBe('font-weight: bold;color: dodgerblue;');
    expect(calls[0][2]).toBe('test info');

    // Warn log
    expect(calls[1][0]).toBe('%c%cwarn');
    expect(calls[1][1]).toBe('font-weight: bold;color: orange;');
    expect(calls[1][2]).toBe('test warn');

    // Error log
    expect(calls[2][0]).toBe('%c%cerror');
    expect(calls[2][1]).toBe('font-weight: bold;color: red;');
    expect(calls[2][2]).toBe('test error');
  });

  test('should create new logger with warn level correctly', () => {
    console.log = vi.fn();
    const newLogger = createLogger({ level: 'warn' });
    newLogger.info('test info');
    newLogger.warn('test warn');
    newLogger.error('test error');

    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(2);

    // Warn log
    expect(calls[0][0]).toBe('%c%cwarn');
    expect(calls[0][1]).toBe('font-weight: bold;color: orange;');
    expect(calls[0][2]).toBe('test warn');

    // Error log
    expect(calls[1][0]).toBe('%c%cerror');
    expect(calls[1][1]).toBe('font-weight: bold;color: red;');
    expect(calls[1][2]).toBe('test error');
  });

  test('should log error with stack correctly', () => {
    console.log = vi.fn();
    logger.error(new Error('this is an error message'));
    expect((console.log as Mock).mock.calls[0][0]).toBe('%c%cerror');
  });

  test('use custom labels if the value is passed', () => {
    console.log = vi.fn();
    const logger = createLogger({
      labels: {
        info: '[ Prefix ] Info',
        warn: '[ Prefix ] Warn',
        ready: '[ Prefix ] Ready',
        success: '[ Prefix ] Success',
      },
    });

    logger.info('this is an info message');
    logger.warn('this is a warn message');
    logger.ready('this is a ready message');
    logger.success('this is a success message');

    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%c[ Prefix ] Info');
    expect(calls[1][0]).toBe('%c%c[ Prefix ] Warn');
    expect(calls[2][0]).toBe('%c%c[ Prefix ] Ready');
    expect(calls[3][0]).toBe('%c%c[ Prefix ] Success');
  });

  test('should handle gradient with empty string', () => {
    console.log = vi.fn();
    logger.greet('');
    expect((console.log as Mock).mock.calls.length).toBe(1);
  });

  test('should handle invalid log level gracefully', () => {
    console.log = vi.fn();
    const logger = createLogger({
      level: 'invalid' as any,
    });

    logger.info('test message');
    logger.debug('test message');

    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%cinfo');
    expect(calls[1][0]).toBe('%c%cdebug');
  });

  test('should handle custom log types', () => {
    console.log = vi.fn();
    const logger = createLogger({
      labels: {
        info: 'CUSTOM_INFO',
      },
    });
    logger.info('test message');

    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toBe('%c%cCUSTOM_INFO');
    expect(calls[0][2]).toBe('test message');
  });

  test('should handle multiple arguments in log message', () => {
    console.log = vi.fn();
    logger.info('message', 'arg1', 'arg2');

    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%cinfo');
    expect(calls[0][2]).toBe('message');
  });

  test('should handle objects in log message', () => {
    console.log = vi.fn();
    logger.info('message', { key: 'value' });

    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%cinfo');
    expect(calls[0][2]).toBe('message');
  });
});

describe('color module', () => {
  test('should handle all color functions', () => {
    console.log = vi.fn();

    const logger = createLogger();
    logger.info('red message');
    logger.warn('orange message');
    logger.success('green message');

    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%cinfo');
    expect(calls[1][0]).toBe('%c%cwarn');
    expect(calls[2][0]).toBe('%c%csuccess');
  });

  test('should handle style combinations', () => {
    console.log = vi.fn();

    const logger = createLogger();
    logger.info(['test', 'font-weight: bold; font-style: italic;']);

    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%cinfo');
    expect(calls[0][1]).toBe('font-weight: bold;color: dodgerblue;');
    expect(calls[0][2]).toBe('test');
    expect(calls[0][3]).toBe('font-weight: bold; font-style: italic;');
  });
});

describe('gradient functionality', () => {
  beforeEach(() => {
    console.log = vi.fn();
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  test('should handle empty string', () => {
    logger.greet('');
    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(1); // One call for empty string
    expect(calls[0][0]).toBe('');
  });

  test('should handle whitespace-only strings', () => {
    logger.greet('   \n   \t   ');
    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(1); // One call for whitespace string
    expect(calls[0][0]).toBe('   \n   \t   ');
  });

  test('should handle normal strings', () => {
    logger.greet('Hello World');
    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(1); // One call for content
    expect(calls[0][0]).toBe('Hello World');
  });

  test('should handle multiline strings', () => {
    logger.greet('Hello\nWorld');
    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(1); // One call for content
    expect(calls[0][0]).toBe('Hello\nWorld');
  });
});

describe('utils module', () => {
  test('should handle empty labels', () => {
    console.log = vi.fn();
    const logger = createLogger({ labels: {} });
    logger.info('test message');
    logger.warn('test message');

    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%cinfo');
    expect(calls[1][0]).toBe('%c%cwarn');
  });

  test('should handle array messages', () => {
    console.log = vi.fn();
    logger.info(['message part 1', 'message part 2']);

    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%cinfo');
    expect(calls[0][2]).toBe('message part 1');
    expect(calls[0][3]).toBe('message part 2');
  });
});

describe('color support detection', () => {
  const originalConsoleLog = console.log;

  beforeEach(() => {
    vi.resetModules();
    console.log = originalConsoleLog;
  });

  test('should cache color support result', () => {
    const supportColor = vi.fn(() => true);
    expect(supportColor()).toBe(true);
    expect(supportColor()).toBe(true);
    expect(supportColor).toHaveBeenCalledTimes(2);
  });

  test('should handle color support detection failure', () => {
    const supportColor = vi.fn(() => false);
    expect(supportColor()).toBe(false);
  });

  afterEach(() => {
    console.log = originalConsoleLog;
    vi.resetModules();
    vi.clearAllMocks();
  });
});

describe('logger configuration', () => {
  test('should handle undefined options', () => {
    console.log = vi.fn();
    const logger = createLogger(undefined);
    logger.info('test');
    expect((console.log as Mock).mock.calls.length).toBe(1);
  });

  test('should handle empty string messages', () => {
    console.log = vi.fn();
    const logger = createLogger();
    logger.info('');
    logger.warn('');
    logger.error('');

    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%cinfo');
    expect(calls[1][0]).toBe('%c%cwarn');
    expect(calls[2][0]).toBe('%c%cerror');
  });

  test('should handle non-string messages', () => {
    console.log = vi.fn();
    const logger = createLogger();
    logger.info(123);
    logger.info(undefined);
    logger.info(null);
    logger.info({ test: 'object' });

    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%cinfo');
    expect(calls[0][2]).toBe('123');
    expect(calls[3][0]).toBe('%c%cinfo');
    expect(calls[3][2]).toBe('[object Object]');
  });
});

describe('label handling', () => {
  test('should handle missing label in log type', () => {
    console.log = vi.fn();
    const logger = createLogger();
    logger.log('test message');
    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should handle custom color functions', () => {
    console.log = vi.fn();
    const logger = createLogger({
      labels: {
        info: 'CUSTOM',
      },
    });
    logger.info(['CUSTOM', 'color: purple;']);

    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%cCUSTOM');
    expect(calls[0][2]).toBe('CUSTOM');
    expect(calls[0][3]).toBe('color: purple;');
  });
});

describe('error handling', () => {
  test('should handle Error objects with custom properties', () => {
    console.log = vi.fn();
    const error = new Error('Custom error');
    logger.error(error);
    const calls = (console.log as Mock).mock.calls;
    expect(calls[0][0]).toBe('%c%cerror');
    expect(calls[0][1]).toBe('font-weight: bold;color: red;');
    expect(calls[0][2]).toContain('Custom error');
  });

  test('should handle non-Error objects in error method', () => {
    console.log = vi.fn();
    logger.error({ message: 'error-like object' });
    logger.error('string error');
    logger.error(123);

    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(3);

    // Error-like object
    expect(calls[0][0]).toBe('%c%cerror');
    expect(calls[0][1]).toBe('font-weight: bold;color: red;');
    expect(calls[0][2]).toBe('[object Object]');

    // String error
    expect(calls[1][0]).toBe('%c%cerror');
    expect(calls[1][1]).toBe('font-weight: bold;color: red;');
    expect(calls[1][2]).toBe('string error');

    // Number error
    expect(calls[2][0]).toBe('%c%cerror');
    expect(calls[2][1]).toBe('font-weight: bold;color: red;');
    expect(calls[2][2]).toBe('123');
  });
});
