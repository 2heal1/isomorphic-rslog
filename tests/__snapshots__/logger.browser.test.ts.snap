// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`label handling > should handle missing label in log type 1`] = `
[
  [
    "test message",
  ],
]
`;
